internal/tuning.go

```diff
@@ -0,0 +1,266 @@
+	// The maximum number of slots that this supplier will ever issue.
+	NumSlots int
+
+	sem *semaphore.Weighted
+}
+
+// NewFixedSizeSlotSupplier creates a new FixedSizeSlotSupplier with the given number of slots.
+func NewFixedSizeSlotSupplier(numSlots int) *FixedSizeSlotSupplier {
+	return &FixedSizeSlotSupplier{
+		NumSlots: numSlots,
+		sem:      semaphore.NewWeighted(int64(numSlots)),
+	}
+}
+
+func (f *FixedSizeSlotSupplier) ReserveSlot(ctx context.Context, reserveCtx SlotReserveContext) (*SlotPermit, error) {
+	err := f.sem.Acquire(ctx, 1)
+	if err != nil {
+		return nil, fmt.Errorf("failed to acquire slot: %w", err)
+	}
+
+	return &SlotPermit{}, nil
+}
+func (f *FixedSizeSlotSupplier) TryReserveSlot(SlotReserveContext) *SlotPermit {
+	if f.sem.TryAcquire(1) {
+		return &SlotPermit{}
+	}
+	return nil
+}
+func (f *FixedSizeSlotSupplier) MarkSlotUsed() {}
+func (f *FixedSizeSlotSupplier) ReleaseSlot() {
+	f.sem.Release(1)
+}
+func (f *FixedSizeSlotSupplier) MaximumSlots() int {
+	return f.NumSlots
+}
+
+type slotReservationData struct {
+	taskQueue string
+}
+
+type slotReserveContextImpl struct {
+	taskQueue   string
+	issuedSlots int
+}
+
+func (s slotReserveContextImpl) TaskQueue() string {
+	return s.taskQueue
+}
+
+func (s slotReserveContextImpl) NumIssuedSlots() int {
+	return s.issuedSlots
+}
+
+type trackingSlotSupplier struct {
+	inner SlotSupplier
+
+	issuedSlotsAtomic atomic.Int32
+	slotsMutex        sync.Mutex
+	// Eventually the map values will be slot info once the API is exposed
+	usedSlots               map[*SlotPermit]struct{}
+	taskSlotsAvailableGauge metrics.Gauge
+	taskSlotsUsedGauge      metrics.Gauge
+}
+
+func newTrackingSlotSupplier(inner SlotSupplier, metricsHandler metrics.Handler) *trackingSlotSupplier {
+	tss := &trackingSlotSupplier{
+		inner:                   inner,
+		usedSlots:               make(map[*SlotPermit]struct{}),
+		taskSlotsAvailableGauge: metricsHandler.Gauge(metrics.WorkerTaskSlotsAvailable),
+		taskSlotsUsedGauge:      metricsHandler.Gauge(metrics.WorkerTaskSlotsUsed),
+	}
+	return tss
+}
+
+func (t *trackingSlotSupplier) ReserveSlot(
+	ctx context.Context,
+	data *slotReservationData,
+) (*SlotPermit, error) {
+	permit, err := t.inner.ReserveSlot(ctx, slotReserveContextImpl{
+		taskQueue:   data.taskQueue,
+		issuedSlots: int(t.issuedSlotsAtomic.Load()),
+	})
+	if err != nil {
+		return nil, err
+	}
+	if permit == nil {
+		return nil, fmt.Errorf("slot supplier returned nil permit")
+	}
+	t.issuedSlotsAtomic.Add(1)
+	t.publishMetrics(false)
+	return permit, nil
+}
+func (t *trackingSlotSupplier) TryReserveSlot(data *slotReservationData) *SlotPermit {
+	permit := t.inner.TryReserveSlot(slotReserveContextImpl{
+		taskQueue:   data.taskQueue,
+		issuedSlots: int(t.issuedSlotsAtomic.Load()),
+	})
+	if permit != nil {
+		t.issuedSlotsAtomic.Add(1)
+		t.publishMetrics(false)
+	}
+	return permit
+}
+func (t *trackingSlotSupplier) MarkSlotUsed(permit *SlotPermit) {
+	if permit == nil {
```

Comment:
Why is a `nil` permit ever being used to reserver a slot?

Reply:
Also changed to panic.